<a href="https://colab.research.google.com/github/James-Hagerman/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Mod_2_Assign_1_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [13]:
### YOUR WORK HERE
import pandas as pd
import numpy as np
from scipy import stats

In [1]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data

--2020-08-10 20:11:35--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data’

imports-85.data     100%[===================>]  25.33K  --.-KB/s    in 0.07s   

2020-08-10 20:11:36 (389 KB/s) - ‘imports-85.data’ saved [25936/25936]



In [2]:
column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
                  'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
                  'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
                  'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
                  'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg',
                  'price']

In [6]:
df = pd.read_csv('imports-85.data', names = column_headers)

df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [8]:
df.replace('?', np.NaN)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3               NaN  alfa-romero  ...       21          27  13495
1            3               NaN  alfa-romero  ...       21          27  16500
2            1               NaN  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [11]:
### YOUR WORK HERE
sample = df.sample(10, random_state = 30)

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [14]:
### YOUR WORK HERE

stats.ttest_1samp(sample['highway-mpg'], .5, nan_policy = 'omit')


Ttest_1sampResult(statistic=14.050858301455087, pvalue=1.98756047234675e-07)

Due to a t-statistic of 14.05, and a p-value of .000000198, we **reject** the null hypothesis that the cars he sells typically have a fuel efficiency of 35-mpg. And we suggest the alternative that 35-mpg is not the mean.


Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [18]:
### YOUR WORK HERE
sample2 = df.sample(100, random_state = 30)

In [19]:
stats.ttest_1samp(sample2['highway-mpg'], .5, nan_policy = 'omit')

Ttest_1sampResult(statistic=45.061541466888855, pvalue=8.84958682057164e-68)

Due to a t-statistic of 45.06, and a p-value of 8.84e-68, we **reject** the null hypothesis that the cars he sells typically have a fuel efficiency of 35-mpg. And we suggest the alternative that 35-mpg is not the mean with 99.99% certainty. 


Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Because having a higher sample size will more accurately reflect the statistics of the population, 

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [20]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-08-10 20:57:48--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-08-10 20:57:48 (286 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [21]:
### YOUR WORK HERE
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers)

print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [29]:
df = df.replace({'y': 1, 'n': 0, '?': np.NaN})

In [30]:
condition1 = df['party'] =='republican'

df_rep = df[condition1]

df_rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [31]:
condition2 = df['party'] == 'democrat'

df_dem = df[condition2]

df_dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [43]:
df_dem['education'].mean()

0.14457831325301204

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [45]:
### YOUR WORK HERE
# Dems == reps
# Dems != reps

stats.ttest_ind(df_rep['education'], df_dem['education'], nan_policy= 'omit')

# due to a T-statistic of 20.5 and a p-value of 1.88e-64 we reject the null 
# hypothesis that dems and republicans 

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE